### PR TITLE
[INLONG-7056][Sort] Adjust sort resources according to data scale

### LINKED ISSUE #7056: [Feature][Sort] Adjust sort resources according to data scale
### Description

Currently, the total amount of resources for the Flink Sort Job comes from the configuration file `flink-sort-plugin.properties`, so all submitted sort jobs will use the same amount of resources. When the data scale is large, the resources will be insufficient. When the data scale is small, the resources are wasted.

```properties
# Flink parallelism
flink.parallelism=1
```

Therefore, dynamically adjusting the number of resources according to the amount of data is one of the urgently needed functions

#### Resource Adaptive Adjustment

Theoretically, the processing performance of Flink can reach about `1000/second/core`, of course, it depends on factors such as state-backed.

Influencing factors:

- Data scale: 
- Storage IO bottleneck: 
  When the performance of a single client connection to external storage becomes a bottleneck, it is a good idea to increase the degree of parallelism or the number of threads
- Transformation computational complexity: 
  In the case of a fixed `LoadNode`, it is a deterministic factor
- Advance factors: 
  core/task manager, parallelism/core, and so on.

### Use case

_No response_

### Are you willing to submit PR?

- [ ] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by PeterZh6**
Fixes #7056 

### Motivation

Currently, the total amount of resources for the Flink Sort Job comes from the configuration file `flink-sort-plugin.properties`, meaning that all submitted sort jobs will use the same amount of resources. When the data scale is large, the resources may be insufficient, and when the data scale is small, the resources may be wasted. Therefore, dynamically adjusting the number of resources according to the amount of data is a critically needed function.

### Modifications

- Before submitting a job to Flink with `org.apache.inlong.manager.plugin.flink.FlinkService#submitJobBySavepoint`, the `org.apache.inlong.manager.plugin.flink.FlinkParallelismOptimizer` will first query the average data volume of from the past hour of  the `DataProxy` module and adjust the parallelism based on this data volume. 
- Meanwhile, this function can be swiched on or off and maxmimum message for one core can be configured in `flink-sort-plugin.properties`
- Moved `OpenApiConstants` to `org.apache.inlong.audit.consts`

### Verifying this change

- [ ] This change is a trivial rework/code cleanup without any test coverage.

- [x] This change is already covered by existing tests, such as:
When creating a stream in Data Ingestion, you can try to make the source data constantly increase and reach a significant amount (approximately more than 2000 per second). Then, resubmit the job. You should notice that the parallelism of the Flink job corresponding to the stream will be larger than the default value of 1. This change will also be reflected in the manager logs.

- [] This change added tests and can be verified as follows:

### CODE REVIEW

File: inlong-manager/manager-plugins/base/src/main/resources/flink-sort-plugin.properties [hunk 1]
```diff
@@ -35,3 +35,7 @@ flink.savepoint.directory=file:///data/inlong-sort/savepoints
 flink.parallelism=1
 # flink stop request drain
 flink.drain=false
+# Flink max data count per core per second
+flink.maxpercore=1000
+# switch on or off dynamic parallelism based on data scale
+flink.dynamicParallelism=true
```
Comment (aloyszhang) [inline inlong-manager/manager-plugins/base/src/main/resources/flink-sort-plugin.properties:41]:
```suggestion
flink.dynamic.parallelism.enable=true
```

File: inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/enums/Constants.java [hunk 2]
```diff
@@ -43,6 +43,10 @@ public class Constants {
 
     public static final String FLINK_VERSION = "flink.version";
 
+    public static final String FLINK_MAXPERCORE = "flink.maxpercore";
+
+    public static final String FLINK_DYNAMIC_PARALLELISM = "flink.dynamicParallelism";
```
Comment (aloyszhang) [inline inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/enums/Constants.java:48]: ref the variables' name in `flink-sort-plugin.properties`

File: inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/dto/FlinkConfig.java [hunk 3]
```diff
@@ -40,4 +40,8 @@ public class FlinkConfig {
     // flink version
     private String version;
 
+    private Integer maxpercore;
+
+    // whether to enable dynamic parallelism
+    private Boolean dynamicParallelism;
```
Comment (aloyszhang) [inline inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/dto/FlinkConfig.java:46]: rename the variables' name in according to `flink-sort-plugin.properties`

File: inlong-audit/audit-common/src/main/java/org/apache/inlong/audit/consts/OpenApiConstants.java [hunk 4]
```diff
@@ -15,7 +15,7 @@
  * limitations under the License.
  */
 
-package org.apache.inlong.audit.config;
+package org.apache.inlong.audit.consts;
```
Comment (aloyszhang) [inline inlong-audit/audit-common/src/main/java/org/apache/inlong/audit/consts/OpenApiConstants.java:18]: It seems this modification is not related to flink job parallelism, maybe we can do this in another pull request.

File: inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+@Slf4j
+@Component
+public class FlinkParallelismOptimizer {
+
+    @Value("${audit.query.url:http://127.0.0.1:10080}")
+    public String auditQueryUrl;
+
+    private static final int MAX_PARALLELISM = 2048;
+    private long maximumMessagePerSecondPerCore = 1000L;
+    private static final int DEFAULT_PARALLELISM = 1;
```
Comment (aloyszhang) [inline inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java:62]: DEFAULT_PARALLELISM should be overwritten by the `flink.parallelism` from `flink-sort-plugin.properties`

File: inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    private static final long DEFAULT_ERROR_DATA_VOLUME = 0L;
+    private static final FlowType DEFAULT_FLOWTYPE = FlowType.OUTPUT;
+    private static final String DEFAULT_AUDIT_TYPE = "DataProxy";
+    private static final String AUDIT_CYCLE_REALTIME = "1";
+    // maxmimum data scale counting range in hours
+    private static final int DATA_SCALE_COUNTING_RANGE_IN_HOURS = 1;
+    private static final String AUDIT_QUERY_DATE_TIME_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSS"; // sample time format:
+                                                                                            // 2024-08-23T22:47:38.866
```
Comment (aloyszhang) [inline inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java:70]:
```suggestion
    ; // sample time format: 2024-08-23T22:47:38.866
    private static final String AUDIT_QUERY_DATE_TIME_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSS"
```

File: inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+            log.info("Retrieved data volume: {}", averageDataVolume);
+        } catch (Exception e) {
+            log.error("Error retrieving data volume: {}", e.getMessage(), e);
+            log.warn("Using default data volume: {}", DEFAULT_ERROR_DATA_VOLUME);
+            averageDataVolume = DEFAULT_ERROR_DATA_VOLUME;
+        }
+        int newParallelism = (int) (averageDataVolume / maximumMessagePerSecondPerCore);
+        newParallelism = Math.max(newParallelism, DEFAULT_PARALLELISM); // Ensure parallelism is at least the default
+                                                                        // value
+        newParallelism = Math.min(newParallelism, MAX_PARALLELISM); // Ensure parallelism is at most MAX_PARALLELISM
```
Comment (aloyszhang) [inline inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java:94]:
```suggestion
         // Ensure parallelism is at least the default value
        newParallelism = Math.max(newParallelism, DEFAULT_PARALLELISM);
         // Ensure parallelism is at most MAX_PARALLELISM
        newParallelism = Math.min(newParallelism, MAX_PARALLELISM);
```

File: inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+
+    private static final String LOGTS_DATE_TIME_FORMAT = "yyyy-MM-dd HH:mm:ss";
+    private static final String TIMEZONE_REGEX = "([+-])(\\d):";
+
+    /**
+     * Calculate recommended parallelism based on maximum message per second per core
+     *
+     * @return Recommended parallelism
+     */
+    public int calculateRecommendedParallelism(List<InlongStreamInfo> streamInfos) {
```
Comment (aloyszhang) [inline inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java:80]: Where is this method invoked?

Comment (PeterZh6) [reply inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java:80]: Thanks a lot for the suggestions! I am sorry that I made a mistaking when submitting code. The missing code of invoking `calculateRecommendedParallelism` in `FlinkService` is now added.

File: inlong-audit/audit-common/src/main/java/org/apache/inlong/audit/consts/OpenApiConstants.java (hunk 4 above)
Comment (PeterZh6) [reply inlong-audit/audit-common/src/main/java/org/apache/inlong/audit/consts/OpenApiConstants.java:18]: This pull request depends on some constants in OpenApiConstants for the first time. However, I encountered some dependency problems when I tried to add the audit-service dependency to the maven-plugin module.  @vernedeng  suggested that this refactor may be helpful to resolve the dependency problem. Refactoring this in another pull request will cause compilation issues and lots of troubles.

File: inlong-manager/manager-plugins/base/src/main/resources/flink-sort-plugin.properties [hunk 9]
```diff
@@ -35,3 +35,7 @@ flink.savepoint.directory=file:///data/inlong-sort/savepoints
 flink.parallelism=1
 # flink stop request drain
 flink.drain=false
+# max data count per core per second
+flink.max.data.percore=1000
```
Comment (aloyszhang) [inline inlong-manager/manager-plugins/base/src/main/resources/flink-sort-plugin.properties:39]:
```suggestion
# max msg rate per core
flink.max.msg.rate.percore=1000
```

File: inlong-audit/audit-common/src/main/java/org/apache/inlong/audit/consts/OpenApiConstants.java (hunk 4 above)
Comment (aloyszhang) [reply inlong-audit/audit-common/src/main/java/org/apache/inlong/audit/consts/OpenApiConstants.java:18]: That makes sense, thanks for your explain.

**Comment by aloyszhang** [comment]
@fuweng11 There are some modifications in the manager module, PTAL

**Review by aloyszhang — DISMISSED** [review]

File: inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+
+    /**
+     * Initialize maximum message per second per core based on configuration
+     *
+     * @param maximumMessagePerSecondPerCore The maximum messages per second per core
+     */
+    public void setMaximumMessagePerSecondPerCore(Integer maximumMessagePerSecondPerCore) {
+        if (maximumMessagePerSecondPerCore == null || maximumMessagePerSecondPerCore <= 0) {
+            log.error("Illegal flink.maxpercore property, must be nonnull and positive, using default value: {}",
+                    maximumMessagePerSecondPerCore);
```
Comment (fuweng11) [inline inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java:105]: The value printed by `maximummessagePerSecondPerCore` is null or less than 0.

File: inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+/**
+ *     This class is used to calculate the recommended parallelism based on the maximum message per second per core.
+ *     The data volume is calculated based on the average data count per hour.
+ *     The data count is retrieved from the inlong audit API.
+ */
+@Slf4j
+@Component
+public class FlinkParallelismOptimizer {
+
+    @Value("${audit.query.url:http://127.0.0.1:10080}")
```
Comment (fuweng11) [inline inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java:57]: Is it certain that this annotation will take effect in plugins?

File: inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+     */
+    public int calculateRecommendedParallelism(List<InlongStreamInfo> streamInfos) {
+        long averageDataVolume;
+        InlongStreamInfo streamInfo = streamInfos.get(0);
+        try {
+            averageDataVolume = getAverageDataVolume(streamInfo);
+            log.info("Retrieved data volume: {}", averageDataVolume);
+        } catch (Exception e) {
+            log.error("Error retrieving data volume: {}", e.getMessage(), e);
+            log.warn("Using default data volume: {}", DEFAULT_ERROR_DATA_VOLUME);
```
Comment (fuweng11) [inline inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java:87]: Please delete one of the warm and error logs.

File: inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java (hunk 11 above)
Comment (PeterZh6) [reply inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java:57]: 9ae2d21d-2ac7-4509-bc34-150e09bef3cc
Yes, as I used `ApplicationContextProvider` to instantiate this Spring bean in `FlinkSerivice`, even though `FlinkService` is not Spring-managed. Attached is the log of `manager`

Comment (fuweng11) [reply inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java:57]: This annotation in the plugins directory will not be scanned during loading.

Comment (PeterZh6) [reply inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java:57]: Could you further explain? I didn't quite get it.

**Comment by fuweng11** [comment]
When a group is a real-time synchronization task, there is no dataproxy present. How to handle this situation.

File: inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java (hunk 11 above)
Comment (PeterZh6) [reply inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java:57]: Or could you suggest a way to read the property? When the Spring profile changes, it is quite clumsy to get the file name of the configuration file. Therefore, I can only think of using the `@value` annotation

Comment (fuweng11) [reply inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java:57]: OK

File: inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+        ZonedDateTime endTime = ZonedDateTime.now(dataZone);
+        ZonedDateTime startTime = endTime.minusHours(DATA_SCALE_COUNTING_RANGE_IN_HOURS);
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern(AUDIT_QUERY_DATE_TIME_FORMAT);
+
+        // counting data volume on with DATA_PROXY_OUTPUT auditId
+        int auditId = AuditIdEnum.getAuditId(DEFAULT_AUDIT_TYPE, DEFAULT_FLOWTYPE).getValue();
+        StringJoiner urlParameters = new StringJoiner(AMPERSAND)
+                .add(PARAMS_START_TIME + EQUAL + startTime.format(formatter))
+                .add(PARAMS_END_TIME + EQUAL + endTime.format(formatter))
+                .add(PARAMS_INLONG_GROUP_Id + EQUAL + streamInfo.getInlongGroupId())
```
Comment (fuweng11) [inline inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java:137]: Please rename `PARAMS_INLONG_GROUP_Id` to `PARAMS_INLONG_GROUP_ID`.

File: inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+                .add(PARAMS_INLONG_STREAM_Id + EQUAL + streamInfo.getInlongStreamId())
```
Comment (fuweng11) [inline inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkParallelismOptimizer.java:138]: Ditto.

**Comment by PeterZh6** [comment]
> When a group is a real-time synchronization task, there is no dataproxy present. How to handle this situation.Is there a default value when dataproxy does not exist？

When no audit data of DataProxy is present, the `entity` should be `null`, causing `parseResponseAndCalculateAverageDataVolume` to return `DEFAULT_ERROR_DATA_VOLUME`, which is `0L`
image

**Comment by PeterZh6** [comment]
9e8f6992-bb82-40f3-99b1-1445cea9f45b
Effect of adjusting parallelism is shown in the picture. (Failure is caused by resource limitation of my computer)

File: inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkService.java [hunk 15]
```diff
@@ -213,6 +218,22 @@ private String submitJobBySavepoint(FlinkInfo flinkInfo, SavepointRestoreSetting
...
+            // Ensure parallelism is at least the default value
+            if (recommendedParallelism < parallelism) {
+                recommendedParallelism = parallelism;
+            }
+            if (recommendedParallelism != parallelism) {
+                log.info("switched to recommended parallelism: {}", recommendedParallelism);
+                parallelism = recommendedParallelism;
+            }
+        }
+        log.info("current parallelism: {}", parallelism);
```
Comment (fuweng11) [inline inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkService.java:236]: Is it reasonable that when audit data cannot be obtained, the `recommendedParallelism` will be 0？

Comment (PeterZh6) [reply inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkService.java:236]: Thanks a lot for your extremely rigorous review. However, I believe with the comparison in the red rectangle, the minimum parallelism is ensured, unless the parallelsim itself in `flink-sort-plugin.properties` is illegal.

image

Comment (fuweng11) [reply inlong-manager/manager-plugins/base/src/main/java/org/apache/inlong/manager/plugin/flink/FlinkService.java:235]: OK

**Review by fuweng11 — APPROVED** [review]

**Review by aloyszhang — APPROVED** [review]
LGTM

**Review by dockerzhang — APPROVED** [review]
